### PR TITLE
Add AMQP example

### PR DESCRIPTION
**Pull request by Alfusainey**
An example demonstrating how the amqp 1.0 binding for cloud events can used. 

The example uses the vertx-proton wrapper. The reason is that the wrapper makes it relatively easy to create amqp clients/servers. Secondly, its a wrapper around qpid-protonj which means that it can be used together with the amqp binding for cloud events. Third, the vertx-proton wrapper is commonly used in IoT setups that leverages AMQP for their messaging.

Signed-off-by: Alfusainey Jallow <alf.jallow@gmail.com>

### CODE REVIEW

File: examples/amqp/pom.xml [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  <dependencies>
+  	 <dependency>
+  	    <groupId>io.cloudevents</groupId>
+  	    <artifactId>cloudevents-amqp-proton</artifactId>
+  	    <version>${project.version}</version>
+  	 </dependency>
+     <dependency>
+     	<groupId>io.vertx</groupId>
+     	<artifactId>vertx-core</artifactId>
+     	<version>3.9.3</version>
```
Comment (slinkydeveloper) [inline examples/amqp/pom.xml:21]: Can you use vert.x 4? If not, no problems.

Also can you make the vertx version a property?

File: examples/amqp/pom.xml [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    <version>2.0.0-SNAPSHOT</version>
+  </parent>
+  <modelVersion>4.0.0</modelVersion>
+
+  <artifactId>cloudevents-amqp-vertx</artifactId>
+
+  <dependencies>
+  	 <dependency>
+  	    <groupId>io.cloudevents</groupId>
+  	    <artifactId>cloudevents-amqp-proton</artifactId>
```
Comment (slinkydeveloper) [inline examples/amqp/pom.xml:15]: s/cloudevents-amqp-proton/cloudevents-amqp-proton-example

File: examples/amqp/src/main/java/io/cloudevents/examples/amqp/vertx/AmqpClient.java [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+        switch (action) {
+
+        case SEND_MESSAGE:
+            connectToServer(vertx, SERVER_HOST, SERVER_PORT)
+                    .compose(conn -> {
+                        connection = conn;
+                        writer.printf("[Client] Connected to %s:%s", SERVER_HOST, SERVER_PORT);
+
+                        return openSenderLink();
+                    }).map(sender -> {
```
Comment (slinkydeveloper) [inline examples/amqp/src/main/java/io/cloudevents/examples/amqp/vertx/AmqpClient.java:61]: in vert.x 4 you can use `onSuccess` if you need to run something only when the future succeed, but without returning anything

File: examples/amqp/src/main/java/io/cloudevents/examples/amqp/vertx/AmqpClient.java [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+                        message.setApplicationProperties(new ApplicationProperties(attributes));
+
+                        sender.send(message, delivery -> {
+                            if (Accepted.class.isInstance(delivery.getRemoteState())) {
+                                writer.println("[Client:] message delivered and accepted by remote peer");
+                            }
+                            connection.close();
+                        });
+                        return null;
+                    }).otherwise(t -> {
```
Comment (slinkydeveloper) [inline examples/amqp/src/main/java/io/cloudevents/examples/amqp/vertx/AmqpClient.java:84]: `onFailure`, similar to `onSuccess` explained above

File: examples/amqp/src/main/java/io/cloudevents/examples/amqp/vertx/AmqpClient.java [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+                        // set attributes
+                        final Map<String, Object> attributes = new HashMap<>();
+                        attributes.put("cloudEvents:type", "com.example.sampletype1");
+                        attributes.put("cloudEvents:source", "http://127.0.0.1/amqp-client");
+                        attributes.put("cloudEvents:id", "client-id");
+                        attributes.put("cloudEvents:specversion", "1.0");
+                        attributes.put("cloudEvents:time", "2020-11-06T21:47:12.037467+00:00");
+                        message.setApplicationProperties(new ApplicationProperties(attributes));
+
+                        sender.send(message, delivery -> {
```
Comment (slinkydeveloper) [inline examples/amqp/src/main/java/io/cloudevents/examples/amqp/vertx/AmqpClient.java:77]: Why are you manually writing the event here?

File: examples/amqp/src/main/java/io/cloudevents/examples/amqp/vertx/AmqpClient.java [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+                        return null;
+                    });
+            break;
+        case RECEIVE_MESSAGE:
+            connectToServer(vertx, SERVER_HOST, SERVER_PORT)
+                    .compose(conn -> {
+                        connection = conn;
+                        writer.println("[Client] Connected");
+                        return Future.succeededFuture();
+                    }).map(success -> {
```
Comment (slinkydeveloper) [inline examples/amqp/src/main/java/io/cloudevents/examples/amqp/vertx/AmqpClient.java:95]: This lambda has only one stmt (the return), collapse it without the {} brackets and return stmt

File: examples/amqp/src/main/java/io/cloudevents/examples/amqp/vertx/AmqpClient.java [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+                        });
+                        return null;
+                    }).otherwise(t -> {
+                        writer.printf("[Client] Connection failed (%s)", t.getCause().getMessage()); 
+                        return null;
+                    });
+            break;
+        case RECEIVE_MESSAGE:
+            connectToServer(vertx, SERVER_HOST, SERVER_PORT)
+                    .compose(conn -> {
```
Comment (slinkydeveloper) [inline examples/amqp/src/main/java/io/cloudevents/examples/amqp/vertx/AmqpClient.java:91]: You could use `onSuccess` here too

File: examples/amqp/src/main/java/io/cloudevents/examples/amqp/vertx/AmqpClient.java [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+            return;
+        }
+
+        final String action = args[0].toLowerCase();
+
+        final Vertx vertx = Vertx.vertx();
+
+        switch (action) {
+
+        case SEND_MESSAGE:
```
Comment (slinkydeveloper) [inline examples/amqp/src/main/java/io/cloudevents/examples/amqp/vertx/AmqpClient.java:54]: To improve the readability of this code, can you put the two cases in two different functions? So here you just do the switch and then for each case, you have the function that executes it

**Review by slinkydeveloper — COMMENTED** [review]
Mostly stylistic comments, looks good!

File: examples/amqp/pom.xml (hunk 1 above)
Comment (Alfusainey) [reply examples/amqp/pom.xml:21]: sure, i can use vert.x 4. the onSuccess call is nice 👍

File: examples/amqp/src/main/java/io/cloudevents/examples/amqp/vertx/AmqpClient.java (hunk 5 above)
Comment (Alfusainey) [reply examples/amqp/src/main/java/io/cloudevents/examples/amqp/vertx/AmqpClient.java:77]: i dont understand. can you please explain what you mean?

i am sending a proton amqp message to the server and adding  the cloudevent attribute names and values as application-properties.

File: examples/amqp/src/main/java/io/cloudevents/examples/amqp/vertx/AmqpClient.java (hunk 8 above)
Comment (Alfusainey) [reply examples/amqp/src/main/java/io/cloudevents/examples/amqp/vertx/AmqpClient.java:54]: yes sure

File: examples/amqp/src/main/java/io/cloudevents/examples/amqp/vertx/AmqpClient.java (hunk 5 above)
Comment (slinkydeveloper) [reply examples/amqp/src/main/java/io/cloudevents/examples/amqp/vertx/AmqpClient.java:77]: Could you create a `CloudEvent` and write it using the message writer more than manually writing the attributes here?

**Review by slinkydeveloper — APPROVED** [review]

**Comment by slinkydeveloper** [comment]
Can you resolve the conflicts so we can merge?

**Comment by Alfusainey** [comment]
> Can you resolve the conflicts so we can merge?

done! also renamed the module name in examples parent module  to `amqp-proton` (instead of  `amqp`)

**Pull request merged**